modules/substrate/src/justification.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Adapted copy of substrate/client/finality-grandpa/src/justification.rs. If origin
+//! will ever be moved to the sp_finality_grandpa, we should reuse that implementation.
+
+// TODO: remove on actual use
+#![allow(dead_code)]
+
+use codec::Decode;
+use finality_grandpa::{voter_set::VoterSet, Chain, Error as GrandpaError};
+use frame_support::RuntimeDebug;
+use sp_finality_grandpa::{AuthorityId, AuthoritySignature, SetId};
+use sp_runtime::traits::Header as HeaderT;
+use sp_std::collections::{btree_map::BTreeMap, btree_set::BTreeSet};
+
+/// Justification verification error.
+#[derive(RuntimeDebug, PartialEq)]
+pub enum Error {
+	/// Failed to decode justification.
+	JustificationDecode,
+	/// Justification is finalizing unexpected header.
+	InvalidJustificationTarget,
+	/// Invalid commit in justification.
+	InvalidJustificationCommit,
+	/// Justification has invalid authority singature.
+	InvalidAuthoritySignature,
+	/// The justification has precommit for the header that has no route from the target header.
+	InvalidPrecommitAncestryProof,
+	/// The justification has 'unused' headers in its precommit ancestries.
+	InvalidPrecommitAncestries,
+}
+
+/// Verify that justification, that is generated by given authority set, finalizes given header.
+pub fn verify_justification<Header: HeaderT>(
```

Comment:
Where would we verify that the `finalized_target` actually builds on top of the latest finalized block we know of? I guess we need the ancestry hashes for that?

Reply:
This is outside of scope of this function. This function is verifying that justification, that is generated by given authority set, finalizes given header (`finalized_target`).